control/grpc.py

```diff
@@ -645,6 +669,56 @@ def get_ns_id_message(self, nsid, uuid):
             ns_id_msg = f"using UUID {uuid} "
         return ns_id_msg
 
+    def set_ana_state(self, request, context=None):
+        return self.execute_grpc_function(self.set_ana_state_safe, request, context)
+
+    def set_ana_state_safe(self, ana_info: pb2.ana_info, context=None):
+        """Sets ana state for this gateway."""
+        self.logger.info(f"Received request to set ana states {ana_info.states}")
+        state = self.gateway_state.local.get_state()
```

Comment:
is "with self.omap_lock(context=context):" really required? looking [here](https://github.com/baum/ceph-nvmeof/blob/d8a1f0b0c788b13049f2b0d3177205c582f7d0b1/control/grpc.py#L237-L246) it appears we're holding the locks in `_safe` function.

Reply:
Again, the _safe function only holds the RPC lock, not the OMAP one.

Reply:
[def execute_grpc_function](https://github.com/baum/ceph-nvmeof/blob/d8a1f0b0c788b13049f2b0d3177205c582f7d0b1/control/grpc.py#L241-L246meof/blob/d8a1f0b0c788b13049f2b0d3177205c582f7d0b1/control/grpc.py#L241-L246)  does not take OMAP lock?

Reply:
No. Just RPC lock:

```
    def _grpc_function_with_lock(self, func, request, context):
        with self.rpc_lock:
            return func(request, context)

    def execute_grpc_function(self, func, request, context):
        return self.omap_lock.execute_omap_locking_function(self._grpc_function_with_lock, func, request, context)
```
We call execute_grpc_function() with a pointer to a _safe function. It passes this to _grpc_function_with_lock() which locks the RPC lock and call the _safe function. Now, within the the _safe function we only lock OMAP if we need to.

Reply:
OK, in principle it's wrong but in the specific case of listeners we can use the local state and not worry about it being out of date as we only care about listeners belonging to our local gateway. So, it could be that some other gateway created a listener which was added to the OMAP file and we don't1 have it in our local state but we would have ignored it anyway. Also, the set ANA state SPDK call doesn't change anything in OMAP so it's OK to do it without holding the lock.